src/frontend/types.ts

```diff
@@ -0,0 +1,297 @@
+ */
+export type ImmutableCast<Type> = RecordOf<{
+  [Property in keyof Type]: Type[Property] extends (infer U)[]
+    ? U extends object
+      ? List<ImmutableCast<U>>
+      : List<U>
+    : Type[Property] extends (infer U)[] | undefined
+    ? U extends object
+      ? List<ImmutableCast<U>> | undefined
+      : List<U> | undefined
+    : Type[Property] extends object
+    ? ImmutableCast<Type[Property]>
+    : Type[Property];
+}>;
+
+export type AppItemTypeRecord = ImmutableCast<AppItemType>;
+export type DocumentItemTypeRecord = ImmutableCast<DocumentItemType>;
+export type FolderItemTypeRecord = ImmutableCast<FolderItemType>;
+export type H5PItemTypeRecord = ImmutableCast<H5PItemType>;
+export type EmbeddedLinkItemTypeRecord = ImmutableCast<EmbeddedLinkItemType>;
+export type LocalFileItemTypeRecord = ImmutableCast<LocalFileItemType>;
+export type S3FileItemTypeRecord = ImmutableCast<S3FileItemType>;
+export type ShortcutItemTypeRecord = ImmutableCast<ShortcutItemType>;
+export type EtherpadItemTypeRecord = ImmutableCast<EtherpadItemType>;
+
+export type ItemRecord =
+  | AppItemTypeRecord
+  | DocumentItemTypeRecord
+  | FolderItemTypeRecord
+  | H5PItemTypeRecord
+  | EmbeddedLinkItemTypeRecord
+  | LocalFileItemTypeRecord
+  | S3FileItemTypeRecord
+  | ShortcutItemTypeRecord
+  | EtherpadItemTypeRecord;
+
+export type EtherpadRecord = RecordOf<Etherpad>;
+
+export type MemberExtraRecord = ImmutableCast<MemberExtra>;
+
+export type MemberRecord = ImmutableCast<Member<MemberExtra>>;
+
+export type ItemMembershipRecord = ImmutableCast<ItemMembership>;
+
+export type ChatMention = {
+  id: string;
+  itemPath: string;
+  message: string;
+  messageId: string;
+  memberId: string;
+  creator: string;
+  createdAt: string;
+  updatedAt: string;
+  status: MentionStatus;
+};
+
+export type ChatMentionRecord = ImmutableCast<ChatMention>;
+
+export type PartialChatMention = {
+  id: string;
+  status: MentionStatus;
+};
+
+export type MemberMentions = {
+  memberId: string;
+  mentions: ChatMention[];
+};
+
+export type MemberMentionsRecord = ImmutableCast<MemberMentions>;
+
+export type MessageBodyType = { message: string; mentions?: string[] };
+
+export type Category = {
+  id: UUID;
+  name: string;
+  type: UUID;
+};
+
+export type CategoryRecord = RecordOf<Category>;
+
+// todo: Do we use this or the one with the `type` prop ??
+export type CategoryType = {
+  id: UUID;
+  name: string;
+};
+
+export type CategoryTypeRecord = RecordOf<CategoryType>;
+
+export type ChatMessage = {
+  id: string;
+  chatId: string;
+  creator: string;
+  createdAt: string;
+  updatedAt: string;
+  body: string;
+};
+
+export type ChatMessageRecord = ImmutableCast<ChatMessage>;
+
+export type ItemChat = {
+  id: string;
+  messages: ChatMessage[];
+};
+
+export type ItemChatRecord = ImmutableCast<ItemChat>;
+
+export type ItemTag = {
+  id: UUID;
+  itemPath: string;
+  tagId: UUID;
+  createdAt: string;
+  creator: string;
+};
```

Comment:
Define in the backend? Probably it already exists.

Reply:
I think we will have to update the plugin because the type defined there is not complete ...